File: research/experiments/sweep/runs/opt-large-python-sonnet45/src/vec_ops.py
"""Vector operations for n-dimensional optimization."""

import math
from typing import List


def dot(a: List[float], b: List[float]) -> float:
    """Dot product of two vectors."""
    return sum(a[i] * b[i] for i in range(len(a)))


def norm(v: List[float]) -> float:
    """Euclidean (L2) norm."""
    return math.sqrt(sum(x * x for x in v))


def norm_inf(v: List[float]) -> float:
    """Infinity norm (max absolute value)."""
    return max(abs(x) for x in v) if v else 0.0


def scale(v: List[float], s: float) -> List[float]:
    """Scalar multiplication."""
    return [x * s for x in v]


def add(a: List[float], b: List[float]) -> List[float]:
    """Element-wise addition."""
    return [a[i] + b[i] for i in range(len(a))]


def sub(a: List[float], b: List[float]) -> List[float]:
    """Element-wise subtraction."""
    return [a[i] - b[i] for i in range(len(a))]


def negate(v: List[float]) -> List[float]:
    """Element-wise negation."""
    return [-x for x in v]


def clone(v: List[float]) -> List[float]:
    """Deep copy of a vector."""
    return v.copy()


def zeros(n: int) -> List[float]:
    """Create a vector of n zeros."""
    return [0.0] * n


def add_scaled(a: List[float], b: List[float], s: float) -> List[float]:
    """Fused operation: a + s*b."""
    return [a[i] + s * b[i] for i in range(len(a))]

File: research/experiments/sweep/runs/opt-large-go-sonnet45/vec_ops.go
package optimization

import "math"

// Dot computes the dot product of two vectors
func Dot(a, b []float64) float64 {
	sum := 0.0
	for i := range a {
		sum += a[i] * b[i]
	}
	return sum
}

// Norm computes the Euclidean (L2) norm of a vector
func Norm(v []float64) float64 {
	return math.Sqrt(Dot(v, v))
}

// NormInf computes the infinity norm (max absolute value) of a vector
func NormInf(v []float64) float64 {
	maxVal := 0.0
	for _, val := range v {
		absVal := math.Abs(val)
		if absVal > maxVal {
			maxVal = absVal
		}
	}
	return maxVal
}

// Scale returns a new vector scaled by scalar s
func Scale(v []float64, s float64) []float64 {
	result := make([]float64, len(v))
	for i, val := range v {
		result[i] = val * s
	}
	return result
}

// Add returns element-wise addition of two vectors
func Add(a, b []float64) []float64 {
	result := make([]float64, len(a))
	for i := range a {
		result[i] = a[i] + b[i]
	}
	return result
}

// Sub returns element-wise subtraction of two vectors (a - b)
func Sub(a, b []float64) []float64 {
	result := make([]float64, len(a))
	for i := range a {
		result[i] = a[i] - b[i]
	}
	return result
}

// Negate returns element-wise negation of a vector
func Negate(v []float64) []float64 {
	return Scale(v, -1.0)
}

// Clone returns a deep copy of a vector
func Clone(v []float64) []float64 {
	return append([]float64(nil), v...)
}

// Zeros returns a vector of n zeros
func Zeros(n int) []float64 {
	return make([]float64, n)
}

// AddScaled returns a + s*b (fused operation)
func AddScaled(a, b []float64, s float64) []float64 {
	result := make([]float64, len(a))
	for i := range a {
		result[i] = a[i] + s*b[i]
	}
	return result
}
